dcore/advection.py

```diff
@@ -122,3 +115,53 @@ def apply(self, x_in, x_out):
         dt = self.state.timestepping.dt
         self.solver.solve()
         x_out.assign(x_in + dt*self.dq)
+
+
+class DGAdvection(Advection):
+
+    def __init__(self, state):
+
+        super(DGAdvection, self).__init__(state)
+
+        dt = state.timestepping.dt
+
+        # by convention the last space is the DG space
+        DGSpace = state.V[-1]
```

Comment:
I would be tempted to redefine the `State` interface.  Currently, you access `state.V` which is an opaque lists of spaces which must conform to an implicit interface (note that this is not the only place where you do this, e.g. you assume that in `LinearAdvection_V3` the DG space is the second entry in `state.V`).

So, `State` could have a method (say):

```python
def get_space(self, name):
    return ...
```

You need to know the ordering (for the mixed function space), but build spaces should build the spaces with a descriptive name (e.g. "temperature", "pressure"), so you can get them again.  The state subclasses can document the available function space names (and say something about them).  This way you have fewer magic numbers.